include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -0,0 +1,835 @@
+                auto __sub_group_local_id = __sub_group.get_local_linear_id();
+
+                auto __group_start_idx =
+                    (__block_num * __max_block_size) + (__g * __inputs_per_sub_group * __num_sub_groups_local);
+
+                std::size_t __elements_in_group =
+                    std::min(__n - __group_start_idx, std::size_t(__num_sub_groups_local * __inputs_per_sub_group));
+                std::uint32_t __active_subgroups =
+                    oneapi::dpl::__internal::__dpl_ceiling_div(__elements_in_group, __inputs_per_sub_group);
+                oneapi::dpl::__internal::__lazy_ctor_storage<_InitValueType> __carry_last;
+
+                // propagate carry in from previous block
+                oneapi::dpl::__internal::__lazy_ctor_storage<_InitValueType> __sub_group_carry;
+
+                // on the first sub-group in a work-group (assuming S subgroups in a work-group):
+                // 1. load S sub-group local carry pfix sums (T0..TS-1) to slm
+                // 2. load 32, 64, 96, etc. TS-1 work-group carry-outs (32 for WG num<32, 64 for WG num<64, etc.),
+                //    and then compute the prefix sum to generate global carry out
+                //    for each WG, i.e., prefix sum on TS-1 carries over all WG.
+                // 3. on each WG select the adjacent neighboring WG carry in
+                // 4. on each WG add the global carry-in to S sub-group local pfix sums to
+                //    get a T-local global carry in
+                // 5. recompute T-local pfix values, add the T-local global carries,
+                //    and then write back the final values to memory
+                if (__sub_group_id == 0)
+                {
+                    // step 1) load to Xe slm the WG-local S prefix sums
+                    //         on WG T-local carries
+                    //            0: T0 carry, 1: T0 + T1 carry, 2: T0 + T1 + T2 carry, ...
+                    //           S: sum(T0 carry...TS carry)
+                    std::uint8_t __iters =
+                        oneapi::dpl::__internal::__dpl_ceiling_div(__active_subgroups, __sub_group_size);
+                    auto __subgroups_before_my_group = __g * __num_sub_groups_local;
+                    std::uint8_t __i = 0;
+                    for (; __i < __iters - 1; __i++)
+                    {
+                        __sub_group_partials[__i * __sub_group_size + __sub_group_local_id] =
+                            __tmp_ptr[__subgroups_before_my_group + __i * __sub_group_size + __sub_group_local_id];
+                    }
+                    if (__i * __sub_group_size + __sub_group_local_id < __active_subgroups)
+                    {
+                        __sub_group_partials[__i * __sub_group_size + __sub_group_local_id] =
+                            __tmp_ptr[__subgroups_before_my_group + __i * __sub_group_size + __sub_group_local_id];
+                    }
```

Comment:
We might want to add a helper for ` __sub_group_size + __sub_group_local_id` and add `__sub_group_size` in each iteration to reuse it throughout this code. Same for `__subgroups_before_my_group + __i * __sub_group_size + __sub_group_local_id`.